src/libmongoc/tests/test-mongoc-client-pool.c

```diff
@@ -41,6 +41,32 @@ test_mongoc_client_pool_try_pop (void)
    mongoc_client_pool_destroy (pool);
 }
 
+static void
+test_mongoc_client_pool_pop_timeout (void)
+{
+   mongoc_client_pool_t *pool;
+   mongoc_client_t *client;
+   mongoc_uri_t *uri;
+   int64_t start;
+   int64_t duration_usec;
+
+   uri = mongoc_uri_new (
+      "mongodb://127.0.0.1/?maxpoolsize=1&waitqueuetimeoutms=2000");
+   pool = mongoc_client_pool_new (uri);
+   client = mongoc_client_pool_pop (pool);
+   BSON_ASSERT (client);
+   start = bson_get_monotonic_time ();
+   BSON_ASSERT (!mongoc_client_pool_pop (pool));
```

Comment:
It looks like this is timing out on macOS in the patch build. I'm not certain why, but my best guess is that it is related to what I noted in mongoc-client-pool.c. If somehow the call to `mongoc_cond_timedwait` was returning (i.e. because condition variable was being signaled), and another iteration of the loop happened, it would sleep for another period of `waitQueueTimeoutMS`. Though, that being said I'm not sure why the condition variable would be getting signaled.

If this persists after adding an expiration time, I'd suggest putting in log statements in `mongoc_client_pool_pop` to identify where it is getting stuck.